src/core/shards/logging.cpp

```diff
@@ -100,30 +103,166 @@ struct Msg : public LoggingBase {
+        _formatted_cache.push_back(fmt::to_string(formatted));
+      }
+    }
+
+    return _formatted_cache;
+  }
+
+protected:
+  void sink_it_(const spdlog::details::log_msg &msg) override {
+    if (msg.level >= min_level_) {
+      _generation++;
+      q_.push_back(spdlog::details::log_msg_buffer{msg});
+    }
+  }
+  void flush_() override {}
+
+private:
+  spdlog::details::circular_q<spdlog::details::log_msg_buffer> q_;
+  spdlog::level::level_enum min_level_;
+
+  size_t _last_generation;
+  size_t _generation;
+  std::vector<std::string> _formatted_cache;
+};
+
+using custom_ringbuffer_sink_mt = custom_ringbuffer_sink<std::mutex>;
+using custom_ringbuffer_sink_st = custom_ringbuffer_sink<spdlog::details::null_mutex>;
+
+struct CaptureLog {
+  static SHTypesInfo inputTypes() { return CoreInfo::NoneType; }
+
+  static SHTypesInfo outputTypes() {
+    // FIXME
+    return CoreInfo::StringType;
+    // return CoreInfo::StringSeqType;
+  }
+
+  static SHParametersInfo parameters() { return _params; }
+
+  void setParam(int index, const SHVar &value) {
+    switch (index) {
+    case 0:
+      _n_items = size_t(value.payload.intValue);
+      break;
+    case 1:
+      _min_level = value.payload.stringValue;
+      break;
+    case 2:
+      _pattern = value.payload.stringValue;
+      break;
+    default:
+      break;
+    }
+  }
+
+  SHVar getParam(int index) {
+    auto res = SHVar();
+    switch (index) {
+    case 0:
+      res.valueType = Int;
+      res.payload.intValue = _n_items;
+      break;
+    case 1:
+      res.valueType = String;
+      res.payload.stringValue = _min_level.c_str();
+      break;
+    case 2:
+      res.valueType = String;
+      res.payload.stringValue = _pattern.c_str();
+      break;
+    default:
+      break;
+    }
+    return res;
+  }
+
+  void warmup(SHContext *context) {
+    auto logger = spdlog::default_logger();
+    assert(logger);
+    auto sink = logger->sinks().at(0);
+    assert(sink);
+    auto dist_sink = std::dynamic_pointer_cast<spdlog::sinks::dist_sink_mt>(sink);
+    if (dist_sink) {
+      _ring = std::make_shared<custom_ringbuffer_sink_mt>(_n_items, spdlog::level::from_str(_min_level));
+      _ring->set_formatter(
+          std::unique_ptr<spdlog::formatter>(new spdlog::pattern_formatter(_pattern, spdlog::pattern_time_type::local)));
+      dist_sink->add_sink(_ring);
+    }
+  }
+
+  void cleanup() {
+    auto logger = spdlog::default_logger();
+    assert(logger);
+    auto sink = logger->sinks().at(0);
+    assert(sink);
+    auto dist_sink = std::dynamic_pointer_cast<spdlog::sinks::dist_sink_mt>(sink);
+    if (dist_sink) {
+      dist_sink->remove_sink(_ring);
+      _ring.reset();
+    }
+  }
+
+  SHVar activate(SHContext *context, const SHVar &input) {
+    if (_ring) {
+      if (_ring->is_dirty()) {
+        auto msgs = _ring->last_formatted();
+
+        _s.clear();
+        for (auto i = msgs.begin(); i != msgs.end(); ++i) {
```

Comment:
HMM I wonder if we should implement it instead using suspends but I can see this being ok too

Reply:
Generates a sequence and uses suspend.